include/boost/graph/rank_spanning_branchings.hpp

```diff
@@ -0,0 +1,1159 @@
+
+     private:
+       const EdgeSet * p_es;
+
+    };
+
+    template<typename Edge, typename WeightMap, typename Compare>
+    struct BranchingEntry
+    {
+      Edge                                                        edge;
+      typename property_traits<WeightMap>::value_type             weight;
+      Compare                                                     compare;
+      unordered_set<Edge>                                         branching;
+      unordered_set<Edge>                                         include_edges;
+      unordered_set<Edge>                                         exclude_edges;
+      BranchingEntry(
+	const typename property_traits<WeightMap>::value_type& w,
+	const Edge& e,
+	Compare& comp,
+	const unordered_set<Edge>& b,
+	const unordered_set<Edge>& include,
+	const unordered_set<Edge>& exclude
+      ) :
+	  edge( e ), weight( w ), compare( comp ),
+	  branching( b ), include_edges( include ),
+	  exclude_edges( exclude ){}
+      bool operator<( BranchingEntry const & rhs ) const
+      { return compare( weight, rhs.weight ); }
+    };
+
+    template<typename WeightMap, typename Edge>
+    typename property_traits<WeightMap>::value_type
+    compute_branching_weight(
+      WeightMap& w,
+      const unordered_set<Edge>& branching
+    )
+    {
+
+       typedef typename property_traits<WeightMap>::value_type weight_t;
+
+       scoped_ptr<weight_t> weight;
+       weight.reset();
+
+       BOOST_FOREACH( const Edge& e, branching )
+       {
+         
+         if( !weight )
+         {
+           weight.reset( new weight_t );
+           *weight = get( w, e );
+         }
+         else
+         {
+	   *weight += get( w, e );
+         }
+       }
+
+       return *weight;
+
+    }
+     
+    template <typename Graph, typename BranchingProcessor, typename IndexMap,
+              typename WeightMap, typename Compare, typename Rank,
+              typename Parent>
+    void 
+    rank_spanning_branchings_impl( const Graph& g,
+			           BranchingProcessor bp,
+                                   IndexMap v_id,
+                                   WeightMap w,
+			           Compare comp,
+                                   Rank rank,
+                                   Parent pred1,
+                                   Parent pred2
+			         )
+    {
+
+      typedef typename graph_traits<Graph>::vertex_descriptor Vertex;
+      typedef typename graph_traits<Graph>::edge_descriptor Edge;
+
+      typedef typename property_traits<WeightMap>::value_type weight_t;
+
+      BOOST_CONCEPT_ASSERT(( VertexAndEdgeListGraphConcept<Graph> ));
+
+      BOOST_CONCEPT_ASSERT(( ReadablePropertyMapConcept<IndexMap, Vertex> ));
+      BOOST_CONCEPT_ASSERT(( ReadablePropertyMapConcept<WeightMap, Edge> ));
+
+      unordered_set<Edge> best_branching, empty_set;
+
+      shared_ptr< std::pair<Edge, weight_t> > p;
+
+      Edge e;
+
+      heap::fibonacci_heap<BranchingEntry<Edge, WeightMap, Compare> > Q;
```

Comment:
In the same tradition of other Boost.Graph algorithms, I would recommend trying to allow the user to choose what type of property map and min heap to use, and provide these choices as defaults.
Some users will have special-case problems where different data structures are better, and in the future there might simply be a better property map, or min heap in general.

You'll have to experiment to find the best way to do it, but in general by making these data structures parameters to the algorithm so that users can pass them in by reference and an overload without those parameters that chooses the defaults. It will require some experimentation and might have to be all or nothing.

The other benefit of letting the client pass in data structures by reference is that the final result of computation is in them for the client to query afterwards if they like. (Think, for example, the open and closed sets in A*.)

Reply:
Thanks for this.  I use two priority queues--one to keep track of branchings in the ranking routine and one to keep track of edge weights in the best and next routines.  The former is just a regular priority queue (PriorityQueue concept), but the latter needs to be mergeable (MergablePriorityQueue concept).  Neither one is "visible" to the user, and the former queue will have had all its entries popped off if all branchings have been found, but I do think it makes sense to allow the user to choose, with the Fibonacci heap as the default.  I'll experiment with this (I think there are some examples in Boost Graph), and with the desirability of distinguishing between the two queues.